components/transferForm.tsx

```diff
@@ -506,12 +510,33 @@ function TransferForm(
         setFormState(values);
         setLoading(true);
         try {
-          let cc = await state.connection.wallet.at(props.address);
+          const cc = await state.connection.wallet.at(props.address);
+          const versioned = VersionedApi(props.contract.version, props.address);
+          const proposals: proposals = { transfers: values.transfers };
 
-          let versioned = VersionedApi(props.contract.version, props.address);
           setTimeoutAndHash(
-            await versioned.submitTxProposals(cc, state.connection, values)
+            await versioned.submitTxProposals(cc, state.connection, proposals)
```

Comment:
Thanks for the PR.

The current implementation in this PR is calling `submitTxProposals` then calling `signProposal` . Therefore, user will need to confirm transactions twice with their logging in wallet, like kukai or temple wallet.

It will be nice to let user make one confirm. To do that, the recommend way is 

1. modify  https://github.com/marigold-dev/tzsafe-ui/blob/main/versioned/interface.ts#L111. add two more options, `isSigning` and `isResolving` on `submitTxProposals` , like

```tsx
abstract submitTxProposals(
    cc: WalletContract,
    t: TezosToolkit,
    proposals: proposals,
    convertTezToMutez?: boolean,
    batch?: WalletOperationBatch
    isSigning?: boolean
    isResolving: boolean
  ): Promise<timeoutAndHash>;
```

2. in its implementation (`versioned/version0_x_x.ts`) of `submitTxPrpopsals`, you can call `signProposal` with args `isSigning` , `isResolving` and `batch`. 

Note that

1.  `0.3.4` supporting “message signing” (proof of event challenge TZIP27), this kind of transaction can not be with other transactions in a proposal. For example, please add “transfer” and “message signing” on the new proposal page and click submit. You will find out you actually create two proposals. So when sign immediately is checked, please make sure these two proposal are all signed.
2. If  `versioned/version0_x_x.ts` is changed, please add test cases here https://github.com/marigold-dev/tzsafe-ui/tree/main/tests/e2eTest
    1. There version `0.1.1`, `0.3.3` and `0.3.4`  are we care more.
    2. The others are nice to have to add the test (but still need to make sure their behavior is expected.).

Reply:
Make sense! I was scared to change the versioned files and need to create another one, but if I understood well these versions are related to the contract, right?

I am just wondering how I can call `signProposal` if that method is getting the proposal by ID. Ok, I can change the `signProposal` to receive directly proposal_id and proposal_content. But how can I get proposal_content before creating the proposal? 🤔

Reply:
> but if I understood well these versions are related to the contract, right?

Correct. It's a contract version. 
The contract's repo: https://github.com/marigold-dev/tzsafe

About the `signProposal`,  the correct implementation way is to pass both proposal_id and proposal_content.
The current  implementation way, using proposal id to lookup proposal content on the contract storage, is risky in some edge case but rarely happen. If the blockchan reorg, users can sign the wrong proposal just by passing proposal id. That is why `sign_proposal` entrypoint require users to pass proposal content again, not just proposal id. User supposes to know which proposal they are signing.  I leave a comment here to say about this issue: https://github.com/marigold-dev/tzsafe-ui/blob/main/versioned/version0_3_3.ts#L196

but I don't expected to correct everything in this PR. The following is the presudocode.  It should work like this:


```
submitTxProposals(
    cc: WalletContract,
    t: TezosToolkit,
    proposals: proposals,
    convertTezToMutez?: boolean,
    batch?: WalletOperationBatch
    isSigning?: boolean
    isResolving: boolean
  ): Promise<timeoutAndHash> {

    proposal_content = https://github.com/marigold-dev/tzsafe-ui/blob/main/versioned/version0_3_3.ts#L44-L178
    const params = cc.methodsObject.create_proposal(proposal_content)
    ....
    if isSigning !== undefined {
      return signProposal(isSigning, isResolving, proposal_Id, proposal_content)
    }
}

signProposal (isSigning, isResolving, proposal_Id, proposal_content?) {
    if proposal_content === undefined {
        // fetch from storage to adopt with current implementation. although, it's not fully correct.
    }
}

```

Reply:
I'd suggest you start from not `0.3.4` , `0.3.4` is complicated then others.

Reply:
Oh, makes sense! I made the change in the `0.3.4`, `0.3.3` and `0.1.1`. Should I do the same with the others?

Also, it's missing the tests, I should write tomorrow.

Reply:
> Should I do the same with the others? :pray:

yes, please.